lib/Dialect/FIRRTL/Export/FIREmitter.cpp

```diff
@@ -0,0 +1,714 @@
+//===- FIREmitter.cpp - FIRRTL dialect to .fir emitter --------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This implements a .fir file emitter.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/FIREmitter.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Translation.h"
+#include "llvm/ADT/APSInt.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "export-firrtl"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Emitter
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// An emitter for FIRRTL dialect operations to .fir output.
+struct Emitter {
+  Emitter(llvm::raw_ostream &os, const FIREmitterOptions &options)
+      : os(os), options(options) {}
+  LogicalResult finalize();
+
+  // Indentation
+  raw_ostream &indent() { return os.indent(currentIndent); }
+  void addIndent() { currentIndent += 2; }
+  void reduceIndent() { currentIndent -= 2; }
```

Comment:
Since `currentIndent` is unsigned, perhaps an assertion to avoid unintended modulo?

Reply:
Good point!